### PR TITLE
Add Alerts and Functions

### PR DESCRIPTION
**Pull request by apmasell**

### CODE REVIEW

File: language.md [hunk 1]
```diff
@@ -268,6 +268,24 @@ names in common between the data going into the join and the input format being
 joined against. If there are, they must be eliminated or renamed using a `Let`
 clause.
 
+These is a final type of olive: one to generate an alert:
```
Comment (callunity) [inline language.md]: These -> There

File: shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/ActionProcessor.java [hunk 2]
```diff
@@ -37,7 +43,83 @@
...
+		private Map<String, String> annotations = new TreeMap<>();
+		private String endsAt;
+		@JsonIgnore
+		private Instant expiryTime;
+		private String generatorURL;
+		private final String id = Long.toString(alertIdGenerator.getAndIncrement());
+		private Map<String, String> labels = new TreeMap<>();
+		private String startsAt;
+
+		public void expriesIn(long ttl) {
```
Comment (callunity) [inline shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/ActionProcessor.java]: expiresIn (typo)

File: shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/ActionProcessor.java [hunk 3]
```diff
@@ -459,12 +543,19 @@ protected boolean check(Action action, Information info) {
 
 	private final Map<Action, Information> actions = new ConcurrentHashMap<>();
 
+	private final AutoLock alertLock = new AutoLock();
+
+	private final Map<Map<String, String>, Alert> alerts = new HashMap<>();
+
+	private final String baseUri;;
```
Comment (callunity) [inline shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/ActionProcessor.java]: Extra `;`

File: shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/MasterRunner.java [hunk 4]
```diff
@@ -33,13 +35,31 @@ private void run() {
...
+						final boolean isDuplicated = actionSink.accept(action, filename, line, column, time);
+						if (isDuplicated) {
+							currentActionDuplicates.incrementAndGet();
+						}
+						return isDuplicated;
+					}
+
+					@Override
+					public boolean accept(String[] labels, String[] annotation, long ttl) throws Exception {
+						final boolean isDuplicated = actionSink.accept(labels, annotation, ttl);
```
Comment (callunity) [inline shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/MasterRunner.java]: `alertSink.accept`?

File: shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/Server.java [hunk 5]
```diff
@@ -486,6 +529,42 @@ private void addJson(String url, Function<ObjectMapper, JsonNode> fetcher) {
 		return functionpRepository.stream();
 	}
 
+	private String localname() {
+		URIBuilder builder = null;
+		final String url = System.getenv("LOCAL_URL");
```
Comment (callunity) [inline shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/Server.java]: `LOCAL_URL` doesn't seem to exist anywhere else in the Shesmu code base; should it be described in the installation docs?

Comment (apmasell) [reply shesmu-server/src/main/java/ca/on/oicr/gsi/shesmu/Server.java]: Added to the README.

**Pull request merged**